File: script/nasaPlaneteScrap.py
import os
import json
import time
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC

options = Options()
options.add_argument("--use-gl=desktop")  # Force l'utilisation de l'accélération matérielle OpenGL
options.add_argument("--enable-webgl")  # Active WebGL
options.headless = False  # N'affiche pas la fenêtre
options.add_argument("--window-size=1920,1080")

# options.add_argument('--disable-gpu')  # Optionnel pour certaines versions de Chrome
# options.add_argument('--disable-extensions')
# options.add_argument('--no-sandbox')
# options.add_argument('--disable-dev-shm-usage')
# options.add_argument('--disable-images') 

driver = webdriver.Chrome(executable_path="/usr/local/bin/chromedriver", options=options)

with open("exoplanets_data.json", "r") as file:
    exoplanets_dict = json.load(file)

images_dir = "img"
os.makedirs(images_dir, exist_ok=True)

for planet_name, planet_url in exoplanets_dict.items():

    try:
        print("---------------------------------------------------")
        print("Status : loading url")
        driver.get(planet_url)
        print("Status : url loaded")
        print("Status : searching iframe")
        iframe = driver.find_element(By.TAG_NAME, "iframe")
        driver.switch_to.frame(iframe)
        print("Status : iframe found")
        print("Status : loading canvas")
        driver.implicitly_wait(2)
        WebDriverWait(driver, 60).until( EC.invisibility_of_element_located((By.CSS_SELECTOR, ".loading")) )
        driver.implicitly_wait(5)
        print("Status : canvas loaded")
        print("Status : remove label and shadow")
        # driver.switch_to.default_content()
        footer_divs = driver.find_elements(By.CSS_SELECTOR, ".footerSVGDiv")
        # footer_divs = WebDriverWait(driver, 20).until(EC.visibility_of_all_elements_located((By.CSS_SELECTOR, ".footerBarRight .footerSVGDiv")) )
        # footer_divs = WebDriverWait(driver, 20).until(EC.visibility_of_all_elements_located((By.CSS_SELECTOR, ".footerBarRight .footerSVGDiv") ))
        if len(footer_divs) >= 2:
            footer_divs[0].click()
            # footer_divs[1].click()
            # switch = footer_divs[1].find_element(By.CSS_SELECTOR, ".switch:first-child")
            # switch.click()

        # driver.switch_to.frame(iframe)
        print("Status : remove ui class")
        driver.execute_script("""
        var uiElements = document.querySelectorAll('.ui');
        uiElements.forEach(function(el) {
            el.style.display = 'none';  
        });
        """)
        print("Status : class removed")

        print("Status : searching canvas")
        filename = os.path.join(images_dir, f"{planet_name.replace('/', '_').replace(' ', '-')}.png")
        canvas = driver.find_element(By.CSS_SELECTOR, "canvas")
        print("Status : canvas find")

        print("Status : screenshot initialize")
        canvas.screenshot(filename)
        driver.implicitly_wait(1)
        print(f"Status : screenshot saved for {planet_name} at {filename}")

    except Exception as e:
        print("Status : error")
        print(e)


driver.quit()
print("All captures are saved.")

File: script/nasaPlaneteScrap.js
const puppeteer = require('puppeteer');
const fs = require('fs');
const path = require('path');


function wait(ms) { return new Promise(resolve => setTimeout(resolve, ms)); }


(async () => 
{
    
    const agents = [
        'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/85.0.4183.121 Safari/537.36',
        'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.3 Safari/605.1.15',
    ];
    
    
    const exoplanetsDict = require('./exoplanets_data.json'); // Assurez-vous que le chemin est correct
    const imagesDir = 'img';
    fs.mkdirSync(imagesDir, { recursive: true });
    
    let i = 0;
    for (let planetName in exoplanetsDict) 
    {
        const planetUrl = exoplanetsDict[planetName];
        try 
        {
            console.log("---------------------------------------------------");
            console.log("Status : loading url");
            const browser = await puppeteer.launch(
            {
                headless: true, // Pour voir ce qui se passe, mettez à true pour mode sans tête
                args: ['--use-gl=desktop', '--enable-webgl', '--window-size=1920,1080']
            });
            const page = await browser.newPage();
            await page.setViewport({ width: 1920, height: 1080 });
            // await page.goto(planetUrl);
            // await page.waitForTimeout(8000);
            await page.setUserAgent(agents[i % 2]);
            await page.goto(planetUrl, { waitUntil: 'domcontentloaded', timeout: 60000 });
            await wait(1000);
            console.log("Status : url loaded");
            
            
            console.log("Status : searching iframe");
            const frameHandle = await page.$('iframe');
            const frame = await frameHandle.contentFrame();
            console.log("Status : iframe found");
            
            console.log("Status : loading canvas");
            await frame.waitForSelector('.loading', { hidden: true });
            console.log("Status : canvas loaded");
            
            console.log("Status : remove label and shadow");
            const footerDivs = await frame.$$('.footerSVGDiv');
            if (footerDivs.length >= 2) {
                await footerDivs[0].click(); // Simulate clicks on specified elements
            }
            
            console.log("Status : remove ui class");
            await frame.evaluate(() => {
                const uiElements = document.querySelectorAll('.ui');
                uiElements.forEach(el => el.style.display = 'none');
            });
            console.log("Status : class removed");
            
            console.log("Status : searching canvas");
            const canvas = await frame.$('canvas');
            console.log("Status : canvas find");
            
            console.log("Status : screenshot initialize");
            const filename = path.join(imagesDir, `${planetName.replace('/', '_').replace(/ /g, '-')}.png`);
            await canvas.screenshot({ path: filename });
            await wait(2000);
            console.log(`Status : screenshot saved for ${planetName} at ${filename}`);
            i++;
            await browser.close();

        } catch (e) 
        {
            await browser.close();
            console.log("Status : error", e);
        }
    }

    console.log("All captures are saved.");
})();
